aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -95,22 +95,70 @@ std::tuple<Tensor, Tensor> slogdet(const Tensor& self) {
   return std::make_tuple(det_sign, abslogdet_val);
 }
 
-Tensor pinverse(const Tensor& self, double rcond) {
-  TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())) && self.dim() >= 2,
-              "pinverse(", self.scalar_type(), "{", self.sizes(), "}): expected a tensor with 2 or more dimensions "
+Tensor linalg_pinv(const Tensor& input, const Tensor& rcond, bool hermitian) {
+  TORCH_CHECK((at::isFloatingType(input.scalar_type()) || at::isComplexType(input.scalar_type())) && input.dim() >= 2,
+              "linalg_pinv(", input.scalar_type(), "{", input.sizes(), "}): expected a tensor with 2 or more dimensions "
               "of floating types");
```

Comment:
"of floating types" is misleading. The requirement is that the tensors have a floating point dtype or a complex dtype. 

Also, is this the right place to valid that input and rcond have the same dtype? 

Also also, what happens when the dtype is bfloat16 or half?

Reply:
Oh wait, I'm mistaken. It should not be that the dtype of input and rcond are the same.

Shouldn't it be that rcond must always be the "value type" of input? That is, if input is double rcond should be double, but if input is complex double then rcond should be double, right?

Reply:
I changed the torch check to accept only float, double, cfloat or cdouble types.

As for the rcond, I don't know actually how we should restrict its types. When only a float is given from Python interface it always gets converted to double in C++ and then we always create a scalar tensor of type double.
If directly a tensor is passed, then all types should be valid that allow multiplication with `max_val` tensor (that can be only of type float or double) and that allow comparison with a tensor of type float or double. So I guess any floating and integer types (so everything that is not complex) should be valid here for any allowed input?

Reply:
I added restriction for rcond type not to accept complex types https://github.com/pytorch/pytorch/pull/48399/commits/4de978601f4e2c63bc9464163e39bd23567a9429
And added tests to show that it works for all other types https://github.com/pytorch/pytorch/pull/48399/commits/4ac1eaf87a7d70e05c0b6d6f12586490cb97d48e

Reply:
Your analysis makes sense. Restricting it to not be complex sounds good.